### ISSUE TITLE
v1.7.9 broke validation of boolean type action inputs

### ISSUE THREAD

**Issue by pbcahill** [rhysd/actionlint#596]
Since the release of `v1.7.9`, it is now throwing the following error when validating an action that has a `boolean` input type:

`Error: .github/workflows/myworkflow.yml:22:15: could not parse action metadata in "/home/runner/work/my-action/my-action": line 4: unexpected key "type" for definition of input "myBooleanInput" [action]`

As soon as I revert to use `v1.7.8` of actionlint, it passes like it did before.

**Comment by muzimuzhi** [comment]
Can you provide an example action?

Also, in action metadata, there is no `inputs.<input_id>.type` property.
https://docs.github.com/en/actions/reference/workflows-and-actions/metadata-syntax#inputs

**Comment by pbcahill** [comment]
I only have a private/internal action with a boolean input. But here is the inputs section from that action where the boolean type is 
```
inputs:
  enableDynamicBadgeNumber:
    description: 'Defaults to false.'
    required: false
    type: boolean
    default: false
```

I'm not sure about that `inputs.<input_id>.type` property you mentioned. But this was working fine in v1.7.8 of actionlint.

**Comment by muzimuzhi** [comment]
> Also, in action metadata, there is no `inputs.<input_id>.type` property.
> https://docs.github.com/en/actions/reference/workflows-and-actions/metadata-syntax#inputs

According to the docs for actions metadata syntax, I think it's an enhancement, not a regression.

The JSON schema for GitHub Actions (https://www.schemastore.org/github-action.json) hosted by schemastore.org doesn't accept `inputs.<input_id>.type` too.

<img width="557" height="266" alt="Image" src="https://github.com/user-attachments/assets/589e8e4a-9426-4037-8dc9-c85626db1cbe" />

(took in VScode, with the [YAML extension](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) by Red Hat enabled, which validates YAML using schemastore.org schemas)

**Comment by pbcahill** [comment]
Hmm interesting. In the official GH documentation, they do list `on.workflow_call.inputs.<input_id>.type` with `boolean` as one of the supported options - https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#onworkflow_callinputsinput_idtype. And searching around, looks like GH added the support for action input types in 2022. So the input type of boolean is supported.

**Comment by muzimuzhi** [comment]
They are different.

- https://docs.github.com/en/actions/reference/workflows-and-actions/metadata-syntax#inputs is the docs for action metadata files named `action.y(a)ml`, and
- https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#onworkflow_callinputsinput_idtype is the docs for workflows, located in `.github/workflows`.

If your YAML file contains something like

```yaml
on:
  workflow_call:
   inputs:
     # ...
```
it's definitely a workflow file. But what your snippet shown in https://github.com/rhysd/actionlint/issues/596#issuecomment-3603309033 looked like an action.

Could you share a simplified while complete action/workflow example?

**Comment by pbcahill** [comment]
This is the test workflow that actionlint is throwing an error on:

```
name: Test Workflow

on:
  workflow_dispatch:

jobs:
  test-run:
    name: Test Run
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@8e8c483db84b4bee98b60c0593521ed34d9990e8 #v6.0.1
      
      - name: Test the action with Dynamic User
        uses: ./
        with:
          enableDynamicBadgeNumber: true
```

The `enableDynamicBadgeNumber` input is the boolean input it is complaining about. This test workflow lives in the same repo as the custom action itself (which is why you see `uses: ./`). The inputs section I shared above is from action.yaml in that same repo. Here it is again:

```
inputs:
  enableDynamicBadgeNumber:
    description: 'Defaults to false.'
    required: false
    type: boolean
    default: false
```

**Comment by muzimuzhi** [comment]
First, the [`jobs.<job_id>.steps[*].uses`](https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#jobsjob_idstepsuses) uses a path

> (The path is) relative (`./`) to the default working directory (`github.workspace`, `$GITHUB_WORKSPACE`).

, so `uses: ./.github/workflows` is needed.

Then, look into the error message, it's indeed complaining about the action **metadata**, so the problem stems from the `action.yaml` file.

> `Error: .github/workflows/myworkflow.yml:22:15: could not parse action metadata in "/home/runner/work/my-action/my-action": line 4: unexpected key "type" for definition of input "myBooleanInput" [action]`

Change it to
```yaml
inputs:
  enableDynamicBadgeNumber:
    description: 'Defaults to false.'
    required: false
    default: 'false'
```
then `actionlint` passes.


A complete example, with the action metadata file relocated to `./.github/actions/my-action/action.yaml`:

```
$ tree .github
.github
├── actions
│   └── my-action
│       └── action.yaml
└── workflows
    └── test-workflow.yml

4 directories, 2 files

$ cat .github/actions/my-action/action.yaml
name: my action
description: 'An example GitHub Action'

inputs:
  enableDynamicBadgeNumber:
    description: 'Defaults to false.'
    required: false
    default: 'false'

runs:
  using: 'composite'
  steps:
    - run: echo "Hello World"
      shell: bash

$ cat .github/workflows/test-workflow.yml
name: Test Workflow

on:
  workflow_dispatch:

jobs:
  test-run:
    name: Test Run
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@8e8c483db84b4bee98b60c0593521ed34d9990e8 #v6.0.1

      - name: Test the action with Dynamic User
        uses: ./.github/actions/my-action
        with:
          enableDynamicBadgeNumber: 'true'

$ actionlint .github/workflows/test-workflow.yml
verbose: Linting .github/workflows/test-workflow.yml
verbose: Using project at /Users/guten/Desktop/actionlint-gh596
verbose: Found 0 parse errors in 0 ms for .github/workflows/test-workflow.yml
verbose: Found total 0 errors in 0 ms for .github/workflows/test-workflow.yml
```

**Comment by pbcahill** [comment]
Our custom action is designed to have that boolean input for a reason. We can't change that. And the input "type" is provided by and officially supported by GitHub, so it's not like it is a hacky solution on our side. Seems like this actionlint needs to be able to support input types.

**Comment by pbcahill** [comment]
And as a reminder, actionlint used to support the input type just fine (or at least not error) in versions 1.7.8 and before. So 1.7.9 had a breaking change in this scenario.

**Comment by muzimuzhi** [comment]
> And the input "type" is provided by and officially supported by GitHub, so it's not like it is a hacky solution on our side.

I doubt that it's officially supported by GitHub. Similar stackexchange Q&A: https://stackoverflow.com/questions/76292948/boolean-input-with-default-value.

GitHub does a loose validation on the action metadata file. For example, it accepts `inputs.<input_id>.descripton` (typo of "description"), see https://github.com/treosh/lighthouse-ci-action/commit/4198ec75af54c544d055c0efad7a45546398a325.

> So 1.7.9 had a breaking change in this scenario.

It's added by 104e931 (test checking unknown keys in inputs in action metadata, 2025-11-12), in supporting

> - Check input deprecation in action by checking [deprecationMessage property](https://docs.github.com/en/actions/reference/workflows-and-actions/metadata-syntax#inputsinput_iddeprecationmessage). Using a deprecated input is reported as error if it is not marked as required. See [the document](https://github.com/rhysd/actionlint/blob/main/docs/checks.md#deprecated-inputs-usage) for more details. ([#580](https://redirect.github.com/rhysd/actionlint/issues/580))
>   ```yaml
>   - uses: reviewdog/action-actionlint@v1
>     with:
>       # ERROR: Using a deprecated input
>       fail_on_error: true
>   ```
> <sub>_https://github.com/rhysd/actionlint/releases/tag/v1.7.9_</sub>

**Comment by rhysd** [comment]
@pbcahill You are mixing inputs for workflow call and inputs for action metadata. As @muzimuzhi said, `type` is not supported in action metadata. Please read the following document again.

https://docs.github.com/en/actions/reference/workflows-and-actions/metadata-syntax#inputs

You [referred](https://github.com/rhysd/actionlint/issues/596#issuecomment-3603580597) the document of inputs of workflow call, which is different from inputs of action metadata.

https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#onworkflow_callinputsinput_idtype

Your mistake was overlooked by actionlint until v1.7.8 because actionlint didn't check illegal keys in action metadata inputs. From v1.7.9 actionlint started to check them correctly. So this is the result of improvements in v1.7.9 as @muzimuzhi said.

And please read the contribution guide. Reproducibility is very important. Showing small snippets and just saying 'hey it doesn't work' is not helpful and such reports usually get low priority.

https://github.com/rhysd/actionlint/blob/main/CONTRIBUTING.md#reporting-an-issue

**Comment by genismoreno** [comment]
@pbcahill I had the same issue but after carefully reading the latest response from @rhysd I do agree that there is no issue at all, but a missconfiguration of the action that was not detected by earlier versions of the linter. Thanks @rhysd for the clarification!

**Comment by pbcahill** [comment]
Ok I understand now, wasn't considering the difference between the workflow call inputs vs metadata. Thank you for looking into this and the very prompt responses. I'll close this Issue now.
